File: src/rust/inu_native/src/io/fetch.rs
//! `fetch`: the one api here that talks to something other than telegram, per `common.d.ts`. The
//! `Response` shape, the `AbortSignal` wiring and the `timeout` are prelude js in `fetch.js`.
//!
//! **There is no http client in this crate and there will not be one**: the app already has a
//! network stack, and both egress rules `common.d.ts` states are decidable only where the
//! connection is made. `PluginFetch` runs the exchange, screening every redirect hop against the
//! grant's domain list and every *resolved* address against the loopback/link-local/private ranges;
//! this module does the pre-flight (scheme, shape, the first url's grant) so a refused call never
//! crosses. The host's check is the authoritative one, being the only side that knows what the name
//! resolved to and where the redirects went.
//!
//! A response body is a file the host wrote, reaching js as a [`crate::io::blob`] app-file backing, so
//! it never touches either heap. A *request* body does cross, bounded by
//! [`crate::io::blob::BUILD_LIMIT_BYTES`]; anything bigger is `uploadFile`'s job.

use std::cell::{Cell, RefCell};
use std::collections::HashMap;
use std::rc::Rc;

use rquickjs::{Ctx, Function, Object, Result as JsResult, Runtime, TypedArray, Value};

use crate::api::json_parse;
use crate::engine::error::{
    check_grant, get_or_create_inu, throw_plugin_error, wire_error_to_js, GrantHost, MATCH_DOMAIN,
};
use crate::io::blob::{export_for_host, mint_app_file, resolve_export, BlobState, BUILD_LIMIT_BYTES};
use crate::tg::rpc::{format_exception, pump_jobs, PendingSettle};

const PRELUDE: &str = include_str!("fetch.js");

/// stand-in for the Kotlin `QuickJs.FetchListener`
pub trait FetchHost {
    /// `None` == accepted, settled later through [`fetch_result`]; `Some` == an error to reject
    /// with, as a bare message or a `P`/`R` wire (never an `E` wire - see `QuickJs.RpcListener`)
    fn send(&self, request_id: i64, url: &str, spec_json: &str, body: Option<&[u8]>) -> Option<String>;

    /// the plugin's `AbortSignal` fired or its `timeout` elapsed: stop, and answer nothing. The
    /// engine has already settled the promise, so a late answer is dropped either way.
    fn abort(&self, request_id: i64);
}

pub struct FetchState {
    host: Rc<dyn FetchHost>,
    grants: Rc<dyn GrantHost>,
    blobs: Rc<BlobState>,
    log: crate::Log,
    next_request_id: Cell<i64>,
    pending: RefCell<HashMap<i64, PendingSettle>>,
}

/// what the url has to be before the host is asked anything: a scheme it will speak, and a host
/// name to check the grant against. The host parses it again with a real url parser - this is the
/// pre-flight, not the authority.
struct Target {
    host: String,
}

/// Extracts the host the grant is checked against, refusing every shape whose host is not what it
/// reads as. The `@` case is the one that matters: `http://allowed.com@127.0.0.1/` connects to
/// 127.0.0.1 while the eye (and a naive parser) reads `allowed.com`, so userinfo is refused
/// outright rather than skipped - nothing here has a use for it.
fn parse_target(url: &str) -> Result<Target, String> {
    let Some((scheme, rest)) = url.split_once("://") else {
        return Err("fetch: the url has no scheme".to_string());
    };
    let scheme = scheme.to_ascii_lowercase();
    if scheme != "http" && scheme != "https" {
        return Err(format!("fetch: '{scheme}' is not a scheme this api speaks"));
    }
    let authority = rest.split(['/', '?', '#']).next().unwrap_or_default();
    if authority.contains('@') {
        return Err("fetch: a url with userinfo in it is refused; put credentials in a header".to_string());
    }
    let host = match authority.strip_prefix('[') {
        // an ipv6 literal keeps its brackets out of the name the grant is matched against
        Some(rest) => match rest.split_once(']') {
            Some((inside, _)) => inside.to_string(),
            None => return Err("fetch: the url has an unterminated ipv6 literal".to_string()),
        },
        None => authority.split(':').next().unwrap_or_default().to_string(),
    };
    // a trailing dot is the same name to dns and a different string to the grant's domain match
    let host = host.trim_end_matches('.').to_ascii_lowercase();
    if host.is_empty() {
        return Err("fetch: the url has no host".to_string());
    }
    Ok(Target { host })
}

/// `string | Uint8Array | Blob` for a request body, read here because only this side can read a
/// blob at all
fn read_body(state: &FetchState, value: &Value<'_>) -> Result<Option<Vec<u8>>, (String, String)> {
    if value.is_undefined() || value.is_null() {
        return Ok(None);
    }
    if let Some(text) = value.as_string() {
        let text = text.to_string().map_err(|e| ("internal".to_string(), format!("fetch: {e:?}")))?;
        return Ok(Some(text.into_bytes()));
    }
    if let Ok(typed) = TypedArray::<u8>::from_value(value.clone()) {
        let Some(bytes) = typed.as_bytes() else {
            return Err(("invalid-argument".to_string(), "fetch: the body array is detached".to_string()));
        };
        return Ok(Some(bytes.to_vec()));
    }
    if let Some(wire) = export_for_host(&state.blobs, value) {
        let id =
            wire.strip_prefix('B').and_then(|rest| rest.split(':').next()).and_then(|id| id.parse().ok()).unwrap_or(0);
        let Some(export) = resolve_export(&state.blobs, id) else {
            return Err(("handle-expired".to_string(), "fetch: the body blob is gone".to_string()));
        };
        if export.len() > BUILD_LIMIT_BYTES {
            return Err((
                "quota-exceeded".to_string(),
                format!(
                    "fetch: a request body is capped at {BUILD_LIMIT_BYTES} bytes; use uploadFile for anything bigger",
                ),
            ));
        }
        let bytes = export
            .read(0, export.len())
            .map_err(|_| ("handle-expired".to_string(), "fetch: the body blob is gone".to_string()))?;
        return Ok(Some(bytes));
    }
    if rquickjs::Class::<crate::io::blob::BlobHandle>::from_value(value).is_ok() {
        return Err(("handle-expired".to_string(), "fetch: the body blob was disposed".to_string()));
    }
    Err(("invalid-argument".to_string(), "fetch: the body must be a string, a Uint8Array or a Blob".to_string()))
}

/// The pre-flight, and the whole of this module's own policy: a scheme it speaks, a host the grant
/// covers, a body it can read. Everything past this is the host's.
fn js_send<'js>(
    ctx: &Ctx<'js>,
    state: &Rc<FetchState>,
    url: String,
    spec_json: String,
    body: Value<'js>,
) -> JsResult<Object<'js>> {
    let target = match parse_target(&url) {
        Ok(target) => target,
        Err(message) => return throw_plugin_error(ctx, "invalid-argument", &message, None, None, None),
    };
    // domain-matched, so `@grant fetch(google.com)` covers `api.google.com` and nothing else. The
    // host re-checks this for every redirect hop, which is the half this side cannot see.
    check_grant(ctx, &state.grants, "fetch", Some(&target.host), MATCH_DOMAIN)?;

    let body = match read_body(state, &body) {
        Ok(body) => body,
        Err((code, message)) => return throw_plugin_error(ctx, &code, &message, None, None, None),
    };

    let request_id = state.next_request_id.get();
    state.next_request_id.set(request_id + 1);
    let (promise, settle) = PendingSettle::new(ctx)?;
    state.pending.borrow_mut().insert(request_id, settle);

    if let Some(err) = state.host.send(request_id, &url, &spec_json, body.as_deref()) {
        if let Some(settle) = state.pending.borrow_mut().remove(&request_id) {
            let value = crate::engine::error::host_error_to_js(ctx, &err)?;
            settle.reject_with_value(ctx, value)?;
        }
    }

    let handle = Object::new(ctx.clone())?;
    handle.set("id", request_id)?;
    handle.set("promise", promise)?;
    Ok(handle)
}

pub fn install_fetch<'js>(
    ctx: &Ctx<'js>,
    host: Rc<dyn FetchHost>,
    grants: Rc<dyn GrantHost>,
    blobs: Rc<BlobState>,
    log: crate::Log,
) -> JsResult<Rc<FetchState>> {
    let state = Rc::new(FetchState {
        host,
        grants,
        blobs,
        log,
        next_request_id: Cell::new(1),
        pending: RefCell::new(HashMap::new()),
    });

    let natives = Object::new(ctx.clone())?;
    {
        let state = state.clone();
        let f = Function::new(ctx.clone(), move |ctx: Ctx<'js>, url: String, spec_json: String, body: Value<'js>| {
            js_send(&ctx, &state, url, spec_json, body)
        })?;
        natives.set("send", f)?;
    }
    {
        let state = state.clone();
        let f = Function::new(ctx.clone(), move |request_id: i64| {
            // the promise is settled by whoever called this; the host is only being told to stop
            state.pending.borrow_mut().remove(&request_id);
            state.host.abort(request_id);
        })?;
        natives.set("abort", f)?;
    }

    // captured at install like `reads.js`'s constructors: what the prelude throws, and the clock it
    // measures `timeout` on, must not be decidable by a plugin reassigning a global
    let plugin_error: Value = get_or_create_inu(ctx)?.get("PluginError")?;
    let timers = Object::new(ctx.clone())?;
    for name in ["setTimeout", "clearTimeout"] {
        let f: Value = ctx.globals().get(name)?;
        timers.set(name, f)?;
    }

    let mut options = rquickjs::context::EvalOptions::default();
    options.filename = Some("<inu:fetch>".to_string());
    let factory: Function = ctx.eval_with_options(PRELUDE, options)?;
    factory.call::<_, ()>((natives, plugin_error, timers))?;
    Ok(state)
}

/// Turns the host's `{path, type}` into the `Blob` the `Response` hands over. The size and mtime
/// are read *here* rather than sent, so the seal [`mint_app_file`] checks every read against is one
/// this side wrote - a number crossing from java and a number `std::fs` reports are two ways of
/// asking the same filesystem and only have to disagree once for every read to say the file is gone.
fn mint_body<'js>(ctx: &Ctx<'js>, body: &Object<'js>) -> JsResult<Value<'js>> {
    let path: String = body.get("path")?;
    let mime: String = body.get("type").unwrap_or_default();
    let path = std::path::PathBuf::from(path);
    let (size, mtime) = match std::fs::metadata(&path) {
        Ok(meta) => (
            meta.len(),
            meta.modified()
                .ok()
                .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
                .map(|d| d.as_millis() as i64)
                .unwrap_or(0),
        ),
        // an empty body the host never had to write; a blob over nothing still answers `size` 0 and
        // reads as the empty string, which is what a 204 should look like
        Err(_) => (0, 0),
    };
    mint_app_file(ctx, &path, size, &mime, None, mtime)
}

/// settles one `fetch`: `J<json>` carrying `{status, statusText, url, headers, body}`, or an error
pub fn fetch_result(
    rt: &Runtime,
    context: &rquickjs::Context,
    state: &Rc<FetchState>,
    request_id: i64,
    result_wire: &str,
) {
    context.with(|ctx| {
        let Some(settle) = state.pending.borrow_mut().remove(&request_id) else {
            return;
        };
        if let Some(built) = wire_error_to_js(&ctx, result_wire) {
            match built {
                Ok(value) => {
                    if settle.reject_with_value(&ctx, value).is_err() {
                        (state.log)(&format!("fetch({request_id}) reject failed: {}", format_exception(&ctx)));
                    }
                }
                Err(e) => {
                    settle.release(&ctx);
                    (state.log)(&format!("fetch({request_id}) error decode failed: {e:?}"));
                }
            }
            return;
        }
        let built = (|| -> JsResult<Value> {
            let json = result_wire
                .strip_prefix('J')
                .ok_or_else(|| rquickjs::Exception::throw_message(&ctx, "fetch: malformed host response"))?;
            let value = json_parse(&ctx, json)?;
            let object = value
                .as_object()
                .cloned()
                .ok_or_else(|| rquickjs::Exception::throw_message(&ctx, "fetch: malformed host response"))?;
            let body: Option<Object> = object.get("body")?;
            let blob = match body {
                Some(body) => mint_body(&ctx, &body)?,
                None => Value::new_null(ctx.clone()),
            };
            object.set("body", blob)?;
            Ok(object.into_value())
        })();
        match built {
            Ok(value) => {
                if settle.resolve_with(&ctx, value).is_err() {
                    (state.log)(&format!("fetch({request_id}) resolve failed: {}", format_exception(&ctx)));
                }
            }
            Err(_) => {
                settle.release(&ctx);
                (state.log)(&format!("fetch({request_id}) bad result wire: {}", format_exception(&ctx)));
            }
        }
    });
    pump_jobs(rt, context, state.log.as_ref());
}

/// releases every `Persistent` GC root this state still owns - same contract as [`crate::tg::rpc::dispose`]
pub fn dispose(context: &rquickjs::Context, state: &Rc<FetchState>) {
    context.with(|ctx| {
        for (_, settle) in state.pending.borrow_mut().drain() {
            settle.release(&ctx);
        }
    });
}

#[cfg(test)]
#[path = "fetch_tests.rs"]
mod tests;

File: src/native/src/js/fetch.js
(natives, PluginError, timers) => {
  const { setTimeout, clearTimeout } = timers



  let getHeaderList
  let freeze

  // names and values are checked natively, so `Headers` and the send path share one set of rules
  class Headers {
    #list = []
    #immutable = false

    static {
      getHeaderList = headers => headers.#list
      freeze = (headers) => {
        headers.#immutable = true
        return headers
      }
    }

    constructor(init) {
      if (init === undefined) return
      if (init === null || (typeof init !== 'object' && typeof init !== 'function')) {
        throw new TypeError('Headers: init must be an object')
      }
      if (#list in init) {
        for (const [name, value] of init.#list) this.#list.push([name, value])
        return
      }
      if (typeof init[Symbol.iterator] === 'function') {
        for (const pair of init) {
          const entry = [...pair]
          if (entry.length !== 2) throw new TypeError('Headers: each pair must be [name, value]')
          this.append(entry[0], entry[1])
        }
        return
      }
      for (const name of Object.keys(init)) {
        const value = init[name]
        if (Array.isArray(value)) {
          for (const one of value) this.append(name, one)
        } else {
          this.append(name, value)
        }
      }
    }

    #checkMutable() {
      if (this.#immutable) throw new TypeError('Headers: these headers are immutable')
    }

    #getValues(name) {
      const values = []
      for (const [key, value] of this.#list) {
        if (key === name) values.push(value)
      }
      return values
    }

    #getSorted() {
      const names = []
      for (const [name] of this.#list) {
        if (!names.includes(name)) names.push(name)
      }
      names.sort()
      const out = []
      for (const name of names) {
        const values = this.#getValues(name)
        if (name === 'set-cookie') {
          for (const value of values) out.push([name, value])
        } else {
          out.push([name, values.join(', ')])
        }
      }
      return out
    }

    append(name, value) {
      const entry = [natives.headerName(name), natives.headerValue(value)]
      this.#checkMutable()
      this.#list.push(entry)
    }

    delete(name) {
      const key = natives.headerName(name)
      this.#checkMutable()
      this.#list = this.#list.filter(([other]) => other !== key)
    }

    get(name) {
      const values = this.#getValues(natives.headerName(name))
      return values.length === 0 ? null : values.join(', ')
    }

    getSetCookie() {
      return this.#getValues('set-cookie')
    }

    has(name) {
      const key = natives.headerName(name)
      return this.#list.some(([other]) => other === key)
    }

    set(name, value) {
      const key = natives.headerName(name)
      const normalized = natives.headerValue(value)
      this.#checkMutable()
      const index = this.#list.findIndex(([other]) => other === key)
      if (index === -1) {
        this.#list.push([key, normalized])
        return
      }
      this.#list[index] = [key, normalized]
      this.#list = this.#list.filter(([other], at) => at <= index || other !== key)
    }

    forEach(callback, thisArg) {
      if (typeof callback !== 'function') throw new TypeError('Headers: forEach needs a function')
      for (const [name, value] of this.#getSorted()) callback.call(thisArg, value, name, this)
    }

    entries() {
      return this.#getSorted().values()
    }

    keys() {
      return this.#getSorted().map(([name]) => name).values()
    }

    values() {
      return this.#getSorted().map(([, value]) => value).values()
    }

    [Symbol.iterator]() {
      return this.entries()
    }
  }
  Object.defineProperty(Headers.prototype, Symbol.toStringTag, { value: 'Headers', configurable: true })

  // the host names each header once, lowercased, with every value it saw
  const createResponseHeaders = (raw) => {
    const headers = new Headers()
    const list = getHeaderList(headers)
    if (raw !== undefined && raw !== null) {
      for (const name of Object.keys(raw)) {
        for (const value of raw[name]) list.push([name, value])
      }
    }
    return freeze(headers)
  }

  const flattenHeaders = (init) => {
    const flat = []
    if (init === undefined || init === null) return flat
    for (const [name, value] of getHeaderList(new Headers(init))) flat.push(name, value)
    return flat
  }

  const bodies = new WeakMap()

  const getResponseBody = (response) => {
    const blob = bodies.get(response)
    if (blob === undefined) throw new TypeError('not a Response')
    return blob
  }

  // the body is a `Blob` over the file the host wrote, so every read here inherits that type's
  // ceilings and its `handle-expired` rather than growing a second set of them
  class Response {
    constructor(raw) {
      bodies.set(this, raw.body)
      Object.defineProperties(this, {
        status: { value: raw.status, enumerable: true },
        statusText: { value: raw.statusText, enumerable: true },
        url: { value: raw.url, enumerable: true },
        ok: { value: raw.status >= 200 && raw.status < 300, enumerable: true },
        headers: { value: createResponseHeaders(raw.headers), enumerable: true },
      })
    }

    blob() {
      return Promise.resolve(getResponseBody(this))
    }

    bytes() {
      return getResponseBody(this).bytes()
    }

    arrayBuffer() {
      return getResponseBody(this).arrayBuffer()
    }

    text() {
      return getResponseBody(this).text()
    }

    json() {
      return getResponseBody(this).text().then(JSON.parse)
    }
  }

  const readInit = (init) => {
    if (init === undefined || init === null) return {}
    if (typeof init !== 'object') throw new PluginError('invalid-argument', 'fetch: the second argument must be an options object')
    return init
  }

  const readTimeout = (init) => {
    const raw = init.timeout
    if (raw === undefined || raw === null) return undefined
    if (typeof raw !== 'number' || !Number.isFinite(raw) || raw <= 0) {
      throw new PluginError('invalid-argument', 'fetch: timeout must be a positive number of milliseconds')
    }
    return raw
  }

  const readSignal = (init) => {
    const signal = init.signal
    if (signal === undefined || signal === null) return undefined
    if (typeof signal.addEventListener !== 'function' || typeof signal.aborted !== 'boolean') {
      throw new PluginError('invalid-argument', 'fetch: signal must be an AbortSignal')
    }
    return signal
  }

  const send = (url, raw) => {
    const init = readInit(raw)
    const timeout = readTimeout(init)
    const signal = readSignal(init)
    const body = init.body

    if (signal !== undefined && signal.aborted) {
      return Promise.reject(new PluginError('aborted', 'the request was aborted'))
    }

    // method, headers and redirect are checked natively: this prelude shares its realm with the
    // plugin, which can reassign `RegExp.prototype.test` or `Array.prototype.push` under any check made here
    const started = natives.send(String(url), init.method, flattenHeaders(init.headers), init.redirect, body)

    return new Promise((resolve, reject) => {
      let settled = false
      let timer

      // The first completion wins. Later responses, errors, or aborts do nothing.
      const finish = (run) => {
        if (settled) return false
        settled = true
        if (timer !== undefined) clearTimeout(timer)
        // eslint-disable-next-line no-use-before-define
        if (signal !== undefined) signal.removeEventListener('abort', onAbort)
        run()
        return true
      }

      const giveUp = (error) => {
        if (finish(() => reject(error))) natives.abort(started.id)
      }

      const onAbort = () => giveUp(new PluginError('aborted', 'the request was aborted'))

      started.promise.then(
        raw => finish(() => resolve(new Response(raw))),
        e => finish(() => reject(e)),
      )

      if (signal !== undefined) signal.addEventListener('abort', onAbort)
      if (timeout !== undefined) {
        timer = setTimeout(
          () => giveUp(new PluginError('timed-out', `the request timed out after ${timeout} ms`)),
          timeout,
        )
      }
    })
  }

  // Reject all failures asynchronously, including grant checks, so callers can always use
  // `fetch(...).catch(...)`.
  const fetch = (url, init) => {
    try {
      return send(url, init)
    } catch (e) {
      return Promise.reject(e)
    }
  }

  // Keep Response private: `dom.d.ts` declares only an interface, not a global constructor.
  Object.defineProperty(globalThis, 'fetch', { value: fetch, writable: true, configurable: true })
  Object.defineProperty(globalThis, 'Headers', { value: Headers, writable: true, configurable: true })
}
